File: Changwhi/2024-07-14-Binary-Tree-Paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        result = []
        temp = []
        def traverse(root: Optional[TreeNode]):
            if not root:
                return
            
            if not root.left and not root.right:
                temp.append(str(root.val))
                result.append("->".join(temp))
                temp.pop()
                return
            
            
            temp.append(str(root.val))
            print(temp)
            traverse(root.left)
            traverse(root.right)
            
            temp.pop()

            
            
        traverse(root)
        return result

File: yoojongho/2024-07-14-binary-tree-paths.java
/**
 * Leetcode
 * problem: 257
 * link: https://leetcode.com/problems/binary-tree-paths/description/
 * tag: String, Backtracking, Tree, Depth-First Search, Binary Tree
 * ---
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 * ---
 * 1. dfs algorithm
 * 2. no child, add to result list
 */
class Solution {
    List<String> result;

    public List<String> binaryTreePaths(TreeNode root) {
        result = new ArrayList<String>();
        helper(root, "");
        return result;
    }

    // method for dfs
    private void helper(TreeNode node, String value) {
        if (value.equals("")) {
            value += node.val;
        } else {
            value += "->" + node.val;
        }

        if (node.left == null && node.right == null) {
            result.add(value);
        } else {
            if (node.left != null) {
                dfs(node.left, value);
            }
            if (node.right != null) {
                dfs(node.right, value);
            }
        }
    }
}